### PR TITLE
feat(diff): show diff statistics

### PR DESCRIPTION
**Pull request by nardoor**
closes rustic-rs/rustic#440

### Description

The proposed code contains:
- a DiffStatistics structure used to hold count of stats
  - this structure offers a few helping functions that will look at a given NodeType to increment the associated counter
  - it has a Display implementation used to display the statistics in STDOUT
- this structure is used in the `diff` function

### Exemple

```bash
rustic --log-level debug -r test-repo --password password diff 8e57051d:config ./config
[INFO] using no config file, none of these exist: /home/nardor/.config/rustic/rustic.toml, /etc/rustic/rustic.toml, ./rustic.toml
[INFO] repository local:test-repo: password is correct.
[INFO] using cache at /home/nardor/.cache/rustic/3797e1fe324e97ac068e303e672da3d34af09cd73888bcfd09538cd72692bf7d
[00:00:00] reading index...               ████████████████████████████████████████          2/2                               
[00:00:00] getting snapshot...            ████████████████████████████████████████          0                                
[INFO] getting snapshot...
[00:00:00] getting snapshot...            ████████████████████████████████████████          0                                 
-    "README.md"
M    "bar"
+    "new_dir"
+    "new_dir/new_file"
-    "services"
-    "services/b2.toml"
-    "services/rclone_ovh-hot-cold.toml"
-    "services/s3_aws.toml"
-    "services/s3_idrive.toml"
-    "services/sftp.toml"
-    "services/sftp_hetzner_sbox.toml"
-    "services/webdav_owncloud_nextcloud.toml"
Files:	1 new,	8 removed, 	1 changed
Dirs:	1 new,	1 removed
Others:	0 new,	0 removed
```

### Testing

I am not sure what tests I can add.
Please let me know if you think of any test about this.

Thanks in advance for any feedback.

### CODE REVIEW

File: src/commands/diff.rs [hunk 1]
```diff
@@ -207,40 +261,53 @@ fn diff(
...
             (Some(i1), Some(i2)) => {
                 let path = &i1.0;
                 let node1 = &i1.1;
                 let node2 = &i2.1;
                 match &node1.node_type {
-                    tpe if tpe != &node2.node_type => println!("T    {path:?}"), // type was changed
+                    tpe if tpe != &node2.node_type => {
+                        // type was changed
+                        println!("T    {path:?}");
+                        diff_statistics.changed_file();
```
Comment (aawsome) [inline src/commands/diff.rs:297]: Maybe make a new type of change here? I would think that "changed file" means a changed file content. Here we have a changed node type (e.g. file changes to dir)..

File: src/commands/diff.rs [hunk 2]
```diff
@@ -207,40 +261,53 @@ fn diff(
...
+                    }
                     NodeType::File if !no_content && !file_identical(path, node1, node2)? => {
                         println!("M    {path:?}");
+                        diff_statistics.changed_file();
                     }
                     NodeType::File if metadata && node1.meta != node2.meta => {
                         println!("U    {path:?}");
+                        diff_statistics.changed_file();
```
Comment (aawsome) [inline src/commands/diff.rs:305]: also here. This is changed metadata, e.g. the the time has changed.

File: src/commands/diff.rs [hunk 3]
```diff
@@ -207,40 +261,53 @@ fn diff(
...
                     }
                     NodeType::Symlink { .. } => {
                         if node1.node_type.to_link() != node1.node_type.to_link() {
                             println!("U    {path:?}");
+                            diff_statistics.changed_file();
```
Comment (aawsome) [inline src/commands/diff.rs:310]: And this is a changed target of a symbolic link

**Review by aawsome — CHANGES_REQUESTED** [review]
Thanks for the PR! I would only request small changes:
- maybe name "file changed" into "content changed"
- and either add more detailed "change types" for the other changes which are currently collected under "file changed"
- or add a single "other changes" to collect all those other changes.

**Comment by aawsome** [comment]
@nardoor Sorry for the late review. About tests, in rustic we only have integration tests (and very few)  so you could try to generate some testdata in a tar-files (or just one tar and modify files by hand), run the command and check the output (maybe using the insta crate). But this is pretty complicated as these tests may be OS-specific. In rustic we are not that strict requesting unit tests, so far.

**Comment by nardoor** [comment]
Hello @aawsome, 

Thanks for taking the time to review the proposed changes. 
I'll gladly apply the review comments asap.

About tests, I might try to add an integration test about this feature, but if I get stuck on some OS specifics as you warned me, I might leave it (for later?).

File: src/commands/diff.rs (hunk 1 above)
Comment (nardoor) [reply src/commands/diff.rs:297]: Why not a "node type" change type

Comment (aawsome) [reply src/commands/diff.rs:297]: Yes, a new type counting how many nodes changes the type. That would be more precise

**Comment by nardoor** [comment]
with new types added:

```
❯ ./target/debug/rustic --log-level debug -r test-repo --password password diff 8e57051d:config ./config  --metadata
[INFO] using no config file, none of these exist: /home/nardor/.config/rustic/rustic.toml, /etc/rustic/rustic.toml, ./rustic.toml
[INFO] repository local:test-repo: password is correct.
[INFO] using cache at /home/nardor/.cache/rustic/3797e1fe324e97ac068e303e672da3d34af09cd73888bcfd09538cd72692bf7d
[00:00:00] reading index...               ████████████████████████████████████████          7/7                                                                                              [00:00:00] getting snapshot...            ████████████████████████████████████████          0                                                                                                [INFO] getting snapshot...
U    "README.md"
-    "bar"
T    "full.toml"
+    "full.toml/coco"
+    "lnsymlink"
U    "local.toml"
+    "new_dir"
-    "services/b2.toml"
-    "services/rclone_ovh-hot-cold.toml"
-    "services/s3_aws.toml"
-    "services/s3_idrive.toml"
-    "services/sftp.toml"
-    "services/sftp_hetzner_sbox.toml"
-    "services/webdav_owncloud_nextcloud.toml"
Files   :	1 new,	8 removed, 	0 changed
Symlinks:	1 new,	0 removed, 	0 updated
Dirs    :	1 new,	0 removed
NodeType:	1 updated
Metadata:	2 updated
```

File: src/commands/diff.rs (hunk 1 above)
Comment (nardoor) [reply src/commands/diff.rs]: I added a NodeType type of change statistics for that

File: src/commands/diff.rs (hunk 3 above)
Comment (nardoor) [reply src/commands/diff.rs]: I added a Symlink change type for that

File: src/commands/diff.rs (hunk 2 above)
Comment (nardoor) [reply src/commands/diff.rs]: Same, I added a `Metadata` change type for that

**Review by aawsome — APPROVED** [review]
LGTM! Thanks again @nardoor 

I removed the unused "use" statement.

**Comment by nardoor** [comment]
Thanks for finishing the work and looking at my PR!

**Pull request merged**